### PR TITLE
PEP 695: Add detection and error reporting for the use of incorrect expressions within the scope of a type parameter and a type alias

### LINKED ISSUE #15238: Support for PEP 695
[PEP 695](https://peps.python.org/pep-0695/) was accepted, and it looks like it will make it into Python 3.12. This PEP adds support for a new type parameter syntax for generic classes and methods. It also adds a new syntax for type aliases (both generic and not).

PEP 695 functionality is implemented in pyright today. It would be great to see it also implemented in mypy and the other major type checkers so we could start to use the new features in type stubs.

Here is a rough task list associated with this work:
- [ ] Add support for `infer_variance` keyword parameter in `TypeVar`; see [this section of PEP](https://peps.python.org/pep-0695/#auto-variance-for-typevar) for details
- [x] Add support for variance inference (similar to variance enforcement in Protocol classes today); see [this section of PEP](https://peps.python.org/pep-0695/#variance-inference) for details
- [x] Add support for new AST nodes for `type` statement; see [this section of PEP](https://peps.python.org/pep-0695/#grammar-changes) for details
- [x] Add support for new AST nodes for generic classes and functions; see [this section of PEP](https://peps.python.org/pep-0695/#grammar-changes) for details
- [x] Add support for evaluating type parameters using the new syntax (including TypeVar bounds and constraints, TypeVarTuple and ParamSpec); see [this section of PEP](https://peps.python.org/pep-0695/#type-parameter-declarations) for details
- [x] Add support for defining type aliases using the new `type` statement (including generic and recursive type definitions); see [this section of PEP](https://peps.python.org/pep-0695/#generic-type-alias) for details
- [x] Add support for defining generic classes and functions with explicit type parameter lists using new syntax
- [x] Add support for functional form of `TypeAliasType`, supported for backward compatibility; see [this section of PEP](https://peps.python.org/pep-0695/#runtime-type-alias-class) for details
- [x] Add detection and error reporting for use of traditional TypeVar, TypeVarTuple, or ParamSpec in conjunction with new syntax; see [this section of PEP](https://peps.python.org/pep-0695/#compatibility-with-traditional-typevars) for details
- [ ] Add detection and error reporting for use of assignment expressions, yield, yield from, and await expressions within the type parameter scope; see [this section of PEP](https://peps.python.org/pep-0695/#type-parameter-scopes) for details

### PR DESCRIPTION
**Pull request by Eclips4**
This fixes part of #15238:
> Add detection and error reporting for use of assignment expressions, yield, yield from, and await expressions within the type parameter scope; see [this section of PEP](https://peps.python.org/pep-0695/#type-parameter-scopes) for details

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
This seems incomplete because it detects these cases only at the top level, not anywhere within the expression. For example, `def f[x: (T + (yield 42))](): pass` should also be flagged.

**Review by sobolevn — COMMENTED** [review]
Yes, you need a recursive visitor for this.

File: mypy/fastparse.py [hunk 1]
```diff
@@ -345,6 +345,21 @@ def is_no_type_check_decorator(expr: ast3.expr) -> bool:
     return False
 
 
+def is_incorrect_expression_used(expr: Any) -> ast3.expr | None:
```
Comment (sobolevn) [inline mypy/fastparse.py:348]: `is_` function returns `ast3.expr | None`, it is better to call it `find_`

File: mypy/fastparse.py [hunk 2]
```diff
@@ -1180,6 +1195,29 @@ def visit_ClassDef(self, n: ast3.ClassDef) -> ClassDef:
         self.class_and_function_stack.pop()
         return cdef
 
+    def validate_type_param(self, type_param: Any) -> None:
```
Comment (sobolevn) [inline mypy/fastparse.py:1198]: It should accept `ast_TypeParam` alias, which would be similar to `ast_TypeAlias` and others.

File: mypy/fastparse.py [hunk 3]
```diff
@@ -345,6 +345,21 @@ def is_no_type_check_decorator(expr: ast3.expr) -> bool:
     return False
 
 
+def is_incorrect_expression_used(expr: Any) -> ast3.expr | None:
+    if not hasattr(expr, "_fields"):
+        return None
+    for attr_name in expr._fields:
```
Comment (sobolevn) [inline mypy/fastparse.py:351]: Question: won't `ast.walk` work here?

File: mypy/fastparse.py (hunk 2 above)
Comment (Eclips4) [reply mypy/fastparse.py:1198]: Actually, there is no `ast.TypeParam` but there is an `ast.TypeVar`. I think you mean it.

File: mypy/fastparse.py (hunk 3 above)
Comment (Eclips4) [reply mypy/fastparse.py:351]: Thanks for the hint! Yes, we can.

File: mypy/fastparse.py [hunk 4]
```diff
@@ -345,6 +347,15 @@ def is_no_type_check_decorator(expr: ast3.expr) -> bool:
     return False
 
 
+def find_incorrect_expression(expr: Any) -> ast3.expr | None:
+    if not hasattr(expr, "_fields"):
+        return None
```
Comment (sobolevn) [inline mypy/fastparse.py:352]: `.bound` can only be `expr | None`: https://github.com/python/typeshed/blob/128d8eb904d045b1faa836a0b82a77d30f853cd7/stdlib/_ast.pyi#L1175

`.value` can only be `expr`: https://github.com/python/typeshed/blob/128d8eb904d045b1faa836a0b82a77d30f853cd7/stdlib/_ast.pyi#L1220

```suggestion
def find_incorrect_expression(expr: ast3.expr | None) -> ast3.expr | None:
    if expr is None:
        return None
```

So, let's adjust the type here to be stricter.

**Review by sobolevn — COMMENTED** [review]
Thank you, one last nit :)

**Review by sobolevn — APPROVED** [review]
Thanks! Congrats on your first mypy PR :)

I will leave it open for a couple of days for others to have a potential feedback.

File: mypy/fastparse.py [hunk 5]
```diff
@@ -345,6 +347,15 @@ def is_no_type_check_decorator(expr: ast3.expr) -> bool:
     return False
 
 
+def find_incorrect_expression(expr: ast3.expr | None) -> ast3.expr | None:
```
Comment (JelleZijlstra) [inline mypy/fastparse.py:350]: This name seems a bit too generic, maybe:
```suggestion
def find_disallowed_expression_in_annotation_scope(expr: ast3.expr | None) -> ast3.expr | None:
```
(will need more changes at the call site)

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**